### DISCUSSION TITLE
Enable source generators to augment XML comments

### DISCUSSION THREAD

**Discussion by stephentoub** [dotnet/csharplang#6014] [Language Ideas] [answered]
It's becoming an increasingly common approach for a source generator to fill in a partial method definition: the user writes a partial method without a body, and the source generator writes one with the body.  However, the source generator is unable to augment the XML comments for that method.  If the user puts their own XML comments on their partial declaration, then if the source generator does as well on its definition, the source generator's XML comments either replace the developers (bad) or are ignored (bad).  There does not appear to be a way for a source generator to _add_ to the comments, e.g. additional remarks or examples or any other content that's actually visible for the user in Visual Studio.

cc: @RikkiGibson

**Comment by RikkiGibson** [comment, marked as answer]
I am actually surprised to hear this. The doc comment processing code is pretty geared around the possibility that documentation for a single symbol may be present across multiple declarations. This is the case for doc comments on types: [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRII5wJZICaoC2EhwECyANHiANQA+AAgEwCMAsAFAD0vABAB5kcQoSgIAngD4AFtgHdho8dK6MAzHwAO4mNigAbPiz4BhPgG8ufG8c2MALHwCyACgCUl232u2AvlwBnFy83IJIYggA1siyEAYGAPYKEeIxUmqaOgh6hsbMZpaBQA=).

I suspect the reason this is lost on methods is somewhat an accident to how the symbols work. [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRII5wJZICaoC2EhwECyANHiANQA+AAgEwCMAsAFCMDMABAAcoCGNigAbPiz4BhPgG8ufZXwD06vgB5kcQoWEBPAHwALbJtU69ho0pW9Bw0RKkAWPgFkAFAEoFKvjtlAF8uUM4uByERMUlpOUVOAKC1DU0kfQQAa2RTCHFxAHsLDOEc2yT7fmjnOPdvHwBuMKA===).

I'll review the doc comments spec when I get a chance but my instinct here is to treat it as a bug-fix level change in the compiler.

cc @jaredpar @dotnet/roslyn-compiler for consideration.

**Reply by jaredpar** [reply]
Did not realize that was the case. Based on your analysis I agree it can be treated mostly as a bug fix in the compiler.

**Reply by RikkiGibson** [reply]
Offline we observed that there are issues in types when the "same" element is present multiple times: [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAHpiACAHgGcBXMSAJwE8A+ACyjMOtokaYOMLk6AU3oBrCq2EAbaQHsOI8ZILoAzCQAOPAC5QI0klhIBhEgG8CJa0Y3oALCQCyACgCUFmySs2AvgX98flIyJR4JJgphAGM5GARFUXCVfHUtXX1DYzNLfF8gA==)

In Quick Info, perhaps, the second remarks tag is "dropped". I suspect that both are present in the compiler API but it's being dropped in Quick Info. Perhaps we should change Quick Info to somehow merge them.

**Reply by stephentoub** [reply]
Yes, if I look at the generated XML file, it contains:
```XML
    <members>
        <member name="T:C">
            <summary>hi</summary>
            <remarks>hello</remarks>
            <remarks>second</remarks>
        </member>
    </members>
```

**Reply by stephentoub** [reply]
But, for completeness, if I move the XML comments to the member instead of the type, only one of them is output:
```XML
    <members>
        <member name="M:C.M">
            <summary>second</summary>
        </member>
    </members>
```

**Reply by jmarolf** [reply]
@RikkiGibson can you file a bug over on https://github.com/dotnet/roslyn/ about this? We've discussed the best way to show long or complex content in QuickInfo in the past and don't have a definitive answer for the best design yet. But this like a reasonable change to make (especially considering source generators) so we combine same-elements and display them using the same heuristics.

**Reply by RikkiGibson** [reply]
Opened dotnet/roslyn#60725

**Reply by stephentoub** [reply]
> I'll review the doc comments spec when I get a chance but my instinct here is to treat it as a bug-fix level change in the compiler.

Should I open an issue then on dotnet/roslyn about fixing the merging of comments on methods?

**Reply by RikkiGibson** [reply]
Yes, let's track and address it separately.

**Reply by stephentoub** [reply]
I opened https://github.com/dotnet/roslyn/issues/60730.

**Reply by ThaDaVos** [reply]
Has anyone seen any update on this?
